solr/core/src/test/org/apache/solr/search/TestThinCache.java

```diff
@@ -26,13 +26,63 @@
 import org.apache.solr.SolrTestCaseJ4;
 import org.apache.solr.metrics.SolrMetricManager;
 import org.apache.solr.metrics.SolrMetricsContext;
+import org.apache.solr.util.EmbeddedSolrServerTestRule;
+import org.apache.solr.util.TestHarness;
 import org.apache.solr.util.stats.MetricUtils;
 import org.junit.BeforeClass;
+import org.junit.ClassRule;
 import org.junit.Test;
 
 /** Test for {@link ThinCache}. */
 public class TestThinCache extends SolrTestCaseJ4 {
 
+  @ClassRule public static EmbeddedSolrServerTestRule solrRule = new EmbeddedSolrServerTestRule();
+  public static final String SOLR_NODE_LEVEL_CACHE_XML =
+      "<solr>\n"
+          + "  <caches>\n"
+          + "    <cache name='myNodeLevelCache'\n"
+          + "      size='10'\n"
+          + "      initialSize='10'\n"
+          + "      />\n"
+          + "    <cache name='myNodeLevelCacheThin'\n"
+          + "      class='solr.ThinCache$NodeLevelCache'\n"
+          + "      size='10'\n"
+          + "      initialSize='10'\n"
+          + "      />\n"
+          + "  </caches>\n"
+          + "</solr>";
+
+  @BeforeClass
+  public static void setupSolrHome() throws Exception {
+    Path home = createTempDir("home");
+    Files.writeString(home.resolve("solr.xml"), SOLR_NODE_LEVEL_CACHE_XML);
+
+    solrRule.startSolr(home);
+
+    Path configSet = createTempDir("configSet");
+    copyMinConf(configSet.toFile());
+    // insert a special filterCache configuration
+    Path solrConfig = configSet.resolve("conf/solrconfig.xml");
+    Files.writeString(
+        solrConfig,
+        Files.readString(solrConfig)
+            .replace(
+                "</config>",
+                "<query>\n"
+                    + "<filterCache\n"
+                    + "      class=\"solr.ThinCache\"\n"
+                    + "      parentCacheName=\"myNodeLevelCacheThin\"\n"
+                    + "      size=\"5\"\n"
+                    + "      initialSize=\"5\"/>\n"
+                    + "</query></config>"));
+
+    solrRule.newCollection().withConfigSet(configSet.toString()).create();
+
+    // legacy; get rid of this someday!
+    h = new TestHarness(solrRule.getCoreContainer());
+    lrf = h.getRequestFactory("/select", 0, 20);
```

Comment:
I need to refresh the vision and discuss/debate it.  Your test here, like most Solr tests, uses `assertU`, `assertQ`, `req`, `doc`, which are implemented today on XML strings and uses TestHarness.  But the vision is to embrace SolrClient.  Nonetheless we need convenience methods to assert via XPath.  I could see a new base test class emerging that has methods named like this and that which are called with the same idioms yet actually have different return types  (definitely not XML strings).